core/src/saros/net/xmpp/filetransfer/XMPPFileTransferManager.java

```diff
@@ -0,0 +1,207 @@
+ *       {@link #setDefaultHandler(Consumer)}
+ *   <li>or otherwise get rejected
+ * </ul>
+ *
+ * <p><b>Outgoing</b> file transfers are possible for Files using {@link #fileSendStart(JID, File,
+ * String)} or InputStreams using {@link #streamSendStart(JID, String, InputStream)}.
+ */
+public class XMPPFileTransferManager {
+  private static final int GLOBAL_SMACK_RESPONSE_TIMEOUT = (int) TimeUnit.MINUTES.toMillis(5);
+
+  private static final Logger log = Logger.getLogger(FileTransferManager.class);
+
+  /** Mapping identifiers of expected transfers with consuming CompletableFuture. */
+  private final ConcurrentMap<String, CompletableFuture<XMPPFileTransferRequest>>
+      expectedTransfers = new ConcurrentHashMap<>();
+
+  private final AtomicReference<Consumer<XMPPFileTransferRequest>> defaultTransferHandler =
+      new AtomicReference<>();
+
+  private final AtomicReference<FileTransferManager> smackTransferManager = new AtomicReference<>();
+
+  private final FileTransferListener smackTransferListener =
+      new FileTransferListener() {
+        @Override
+        public void fileTransferRequest(FileTransferRequest request) {
+          XMPPContact contact = contactsService.getContact(request.getRequestor()).orElse(null);
+          if (contact == null) {
+            log.info("received transfer request but contact unknown " + request.getRequestor());
+            return;
+          }
+
+          String identifier = request.getDescription();
+          log.info(
+              "received transfer request, name: "
+                  + request.getFileName()
+                  + ", identifier: "
+                  + identifier
+                  + " by "
+                  + contact);
+
+          if (identifier != null) {
+            CompletableFuture<XMPPFileTransferRequest> future =
+                expectedTransfers.remove(identifier);
+            if (future != null) {
+              future.complete(new XMPPFileTransferRequest(contact, request));
+              return;
+            }
+          }
+
+          Consumer<XMPPFileTransferRequest> defaultListener = defaultTransferHandler.get();
+          if (defaultListener == null) {
+            request.reject();
+          } else {
+            defaultListener.accept(new XMPPFileTransferRequest(contact, request));
+          }
+        }
+      };
+
+  private final IConnectionListener connectionListener =
+      (connection, state) -> {
+        FileTransferManager currentManager = smackTransferManager.getAndSet(null);
+        if (currentManager != null) {
+          currentManager.removeFileTransferListener(smackTransferListener);
+        }
+
+        if (state == ConnectionState.CONNECTED) {
+          FileTransferManager newManager = new FileTransferManager(connection);
+          smackTransferManager.set(newManager);
+          newManager.addFileTransferListener(smackTransferListener);
+        }
```

Comment:
Ok this is probably a matter of taste. I think the solution works good in terms of readability, and adding checks to avoid a single set operation is not necessary.